feature/GOL-1-add-unit-tests-buttons

GUIDOURI opened the pull request:
###  What was done?

- Created unit tests for all button variants (Filled, Elevated, Tonal, Outlined, Text) using React Testing Library.
- Covered edge cases and state changes (e.g., disabled, loading, with icons).
- Integrated animations and design tokens.

###  Why?

To ensure UI components follow expected behavior and appearance under different props and states, increasing maintainability and trust.

###  How to test?

   ````markdown
1. Run:
   ```bash
   pnpm test src/components/button
2. All button variants should be tested successfully.

3. Storybook renders should also match the tested components.

ougarte committed:
chore: initial solution

ougarte committed:
dirty

ougarte committed:
chore(ui): initial ui application multi-platform

GUIDOURI committed:
WIP: progreso inicial de la librería de botones Material 3

GUIDOURI committed:
Test(button): add unit tests for Material 3 buttons (filled, elevated, tonal, outlined, text, icon)

Includes:
- Snapshot-free tests for each variant
- Disabled state test
- Text rendering test
- Ensures accessibility with aria-label for icon buttons

Next steps:
- Add tests for FAB and Segmented buttons
- Test interactions like onClick and loading states

GUIDOURI committed:
chore(test): fix ESLint issues and add Jest testing setup for GOL-1
- Applied ESLint fixes to all button-related test files under __tests__.
- Added "test" script using Jest to package.json.
- Installed @testing-library/react and @testing-library/jest-dom as devDependencies.

GUIDOURI committed:
refactor(button): unify all button variants into a single component

- Consolidated button components (filled, elevated, tonal, outlined, text, icon, segmented) into `button.tsx`
- Preserved variant-specific behavior via optional wrappers (e.g., FilledButton, IconButton)
- Ensures adherence to SOLID principles and Clean Code guidelines
- Removes duplication and centralizes variant handling in the base `Button` component

GUIDOURI committed:
refactor(button): unify all button variants into a single component

- Consolidated button components (filled, elevated, tonal, outlined, text, icon, segmented) into `button.tsx`
- Preserved variant-specific behavior via optional wrappers (e.g., FilledButton, IconButton)
- Ensures adherence to SOLID principles and Clean Code guidelines
- Removes duplication and centralizes variant handling in the base `Button` component

GUIDOURI committed:
refactor(app): move AppController files into 'controller' folder

- Relocated app.controller.ts and app.controller.spec.ts to app/controller/ for better modular organization
- Aligns with recommended NestJS structure and improves maintainability
- Updated import paths accordingly (if needed)

GUIDOURI committed:
test(button): unify all variant tests into a single file

- Consolidated tests for all button variants into `button.test.tsx`
- Preserves coverage for Filled, Elevated, Tonal, Outlined, Text, and Icon buttons
- Simplifies test management and aligns with clean architecture practices

GUIDOURI committed:
test(api): standardize controller and service tests

Refactored AppController and AppService unit tests to follow consistent test naming and structure.
- Removed unnecessary describe blocks
- Applied Arrange-Act-Assert format
- Renamed tests to 'should get a message' for clarity

GUIDOURI committed:
refactor(ui/button): simplify button types by removing unused IconProperties and TextProperties

Updated the Properties type to extend ButtonHTMLAttributes directly and include relevant fields like `as`, `className`, and `children`. This cleanup improves type clarity and maintains full component functionality.

GUIDOURI committed:
fix(ui/button): resolve ESLint and TS issues in button.stories.tsx

Updated import path to use the correct `button` file and added explicit typing to all `args` parameters to prevent implicit any errors. This ensures Storybook runs without TS or ESLint complaints.

GUIDOURI committed:
refactor(ui-button): restructure button.module.css for scoped styles and maintainability

- Rename .buttonBase to .button
- Remove unused .fullWidth utility class
- Scope .icon style to .button .icon
- Nest all variant classes under .button (e.g., .button.filled)
- Move .fadeIn animation to animations.css for separation of concerns

GUIDOURI committed:
feat(ui/button): restore original button styles with Material 3 variants

This commit restores the original button styles using Material Design 3 guidelines.
It includes `.buttonBase` as the base class and supports full variant styles like
.filled, .elevated, .tonal, .outlined, and .text, along with a .fadeIn animation.
Also includes `.icon` and `.fullWidth` support for layout control.

GUIDOURI committed:
fix(ui/button): correct ButtonProperties type export from types

Fixes incorrect type re-export that caused TS error due to missing type in ./button. Now correctly exports from types module.

ougarte commented on apps/ui/src/components/button/button.tsx:
```diff
@@ -0,0 +1,86 @@
+export default Button;
+
+
+export const FilledButton = (properties: Omit<ButtonBaseProperties, 'variant'>) => (
```
I recommend removing all variants `FilledButton | ElevatedButton | TonalButton  | OutlinedButton  | TextButton  | IconButton  | SegmentedButton`

ougarte commented on apps/ui/src/components/button/index.ts:
```diff
@@ -0,0 +1,16 @@
+	default as ButtonOutlinedComponent,
+	default as ButtonFABComponent,
+	default as ButtonSegmentedComponent,
+} from './button'; 
```
I recommend exporting a unique Button `export { default as Button } from './button'`

ougarte commented on apps/ui/src/components/button/styles/animations.css:
```diff
@@ -0,0 +1,21 @@
+@import url('./variables.css');
+
+.button {
```
I recommend moving this style to `'core.css'`

ougarte commented on apps/ui/src/components/button/styles/button.module.css:
```diff
@@ -0,0 +1,94 @@
+@import './animations.css';
+
+/* Estilo base del botón */
+.buttonBase {
```
I recommend updating as `.button`

ougarte commented on apps/ui/src/components/button/styles/button.module.css:
```diff
@@ -0,0 +1,94 @@
+}
+
+/* Estilo para los íconos dentro del botón */
+.icon {
```
I recommend adding a namespace `.button .icon`

ougarte commented on apps/ui/src/components/button/styles/button.module.css:
```diff
@@ -0,0 +1,94 @@
+/* Variantes de Material 3 */
+
+/* Estilo para el botón lleno */
+.filled {
```
I recommend moving into `.button` scope

```css
.button {
    &.filled {}
    /* add more styles here... */
}
```

ougarte commented on apps/ui/src/components/button/button.tsx:
```diff
@@ -0,0 +1,86 @@
+import React from 'react';
+import clsx from 'clsx';
+import styles from './styles/button.module.css';
```
I recommend removing `css modules`

ougarte commented on apps/ui/src/components/button/styles/variants.css:
```diff
@@ -0,0 +1,21 @@
+	/* By default is `text` button */
+
+	&.filled {
+		background-color: transparent;
```
I recommend moving `background-color: transparent;` as 
```css
.button {
    background-color: transparent;
}
```

GUIDOURI committed:
refactor(button): simplify index.ts exports to a single unified Button export

GUIDOURI committed:
refactor(button): extract variants into variants.css and apply consistent button class scoping

GUIDOURI committed:
refactor(animations): move background-color: transparent to .button for better CSS organization

GUIDOURI committed:
style(button): namespace icon class and scope variants under .button as per modular CSS practices

GUIDOURI committed:
refactor(button): remove variant-specific subcomponents and keep only the main Button component

ougarte commented on apps/ui/src/components/button/helpers.tsx:
```diff
@@ -0,0 +1,15 @@
+ * @param properties
+ * @returns
+ */
+// Después
```
I recommend using English comments

ougarte commented on apps/ui/src/components/button/__tests__/button.test.tsx:
```diff
@@ -0,0 +1,85 @@
+	IconButton,
+} from '../button';
+
+describe('Button Variants', () => {
```
I recommend organizing the unit test as follow:

```ts
describe('components/button', () => {
    describe('layout', () => {
        test('should render with default properties', () => {});
        test('should render filled default with default properties', () => {});
        test('should render elevated  default with default properties', () => {});
        // and so on...
    });
});
```

ougarte commented on apps/ui/src/components/button/button.tsx:
```diff
@@ -0,0 +1,41 @@
+}) => {
+	const classes = clsx(
+		'button',
+		`button--${variant}`,
```
I recommend avoiding mixing CSS approaches... so that you could use as follow

```ts
clsx('button', variant, fullWidth, selected, className);
```

ougarte commented on apps/ui/src/components/button/button.tsx:
```diff
@@ -0,0 +1,41 @@
+  className?: string;
+}
+
+const Button: React.FC<ButtonBaseProperties> = ({
```
I recommend Funcional Components

```tsx
export default function Button(properties?: ButtonBaseProperties) {
    // code goes here...
}
```

ougarte commented on apps/ui/src/components/button/__tests__/button.test.tsx:
```diff
@@ -0,0 +1,85 @@
+	OutlinedButton,
+	TextButton,
+	IconButton,
+} from '../button';
```
Quick question: where are these buttons being exported?

ougarte commented on apps/ui/src/components/button/button.tsx:
```diff
@@ -0,0 +1,41 @@
+	icon,
+	selected = false,
+	className,
+	...properties
```
I recommend wrapping `properties` with `toDefaults` helper

ougarte commented on apps/ui/src/components/button/styles/button.module.css:
```diff
@@ -0,0 +1,92 @@
+}
+
+
+.button--fullWidth {
```
I recommend moving to .button as follow

```css
.button {
    &.full-width { }
    .icon {}

    /* and more nested selectors */
}
```

ougarte commented on apps/ui/src/components/button/styles/button.module.css:
I recommend moving all styles to their respective `css` file, and then delete this `button.module.css` file

GUIDOURI committed:
refactor(ui): restructure Button component with modular variants and add unit tests

- Moved Button variants (`filled`, `elevated`, `tonal`, `outlined`, `text`) to individual files under `variants/`
- Separated CSS styles for each variant under `styles/variants/`
- Added `core.css` and `animations.css` for shared styles and transitions
- Applied CSS architecture principles: BEM, OOCSS, SMACSS

test(ui): set up unit testing with Jest and React Testing Library

- Installed `@testing-library/react` and `@testing-library/jest-dom`
- Added `jest.setup.ts` and configured `jest.config.ts` for jsdom environment
- Created `button.test.tsx` with 6 passing tests:
  - renders default (filled) variant
  - renders specified variant
  - applies custom className
  - renders icon if provided
  - handles layout prop
  - applies fullWidth class when true

ougarte commented on apps/ui/src/components/button/__tests__/button.test.tsx:
```diff
@@ -0,0 +1,40 @@
+import { render, screen } from '@testing-library/react';
+import Button from '../button';
+
+describe('Button', () => {
```
I recommend updating as `components/button`

ougarte commented on apps/ui/src/components/button/__tests__/button.test.tsx:
```diff
@@ -0,0 +1,40 @@
+import Button from '../button';
+
+describe('Button', () => {
+	it('renders with default variant (filled)', () => {
```
I recommend using `should` on **all unit-test** (ex: `should render by default`)

ougarte commented on apps/ui/src/components/button/button.tsx:
```diff
@@ -0,0 +1,64 @@
+  icon?: React.ReactNode;
+  selected?: boolean;
+  className?: string;
+  layout?: 'horizontal' | 'vertical' | 'centered';
```
I recommend changing `layout` by `direction` or `alignment`

ougarte commented on apps/ui/src/components/button/button.tsx:
```diff
@@ -0,0 +1,64 @@
+  layout?: 'horizontal' | 'vertical' | 'centered';
+}
+
+const variantClassMap: Record<ButtonVariant, () => string> = {
```
I think this variable is not necesary

ougarte commented on apps/ui/src/components/button/button.tsx:
```diff
@@ -0,0 +1,64 @@
+		layout && `layout--${layout}`,
+		'fade-in',
+		className
+	);
```
I recommend changing as follow:

```tsx
clsx('button fade-in', variant, className, layout, {
    'button--fullWidth': fullWidth
    'button--selected': selected,
});
```

GUIDOURI committed:
refactor(button): remove variantClassMap and simplify variant styles

- Removed variantClassMap logic and variant .tsx files for simplicity
- Consolidated CSS imports directly in button.tsx
- Followed BEM and OOCSS principles for consistent and maintainable styles
- Improved clarity for unit testing and reduced file complexity

GUIDOURI committed:
feat(ui/button): implement reusable Button component with Material 3 variants and full HTML form support

- Created variants: 'filled', 'elevated', 'tonal', 'outlined', 'text' following Material 3 specs
- Modular CSS with BEM + OOCSS + SMACSS, and transition-based animations
- Supports HTML form attributes:
  - type="button" | "submit" | "reset"
  - disabled (including visual state and pointer lock)
  - name, value (for form data submission)
  - aria-pressed (for toggle button accessibility)
- Accepts children, optional icon, fullWidth, layout, and selected props
- Compatible with forms (submit/reset), dialogs, toolbars, FABs, and navigation patterns
- Added interactive states: hover, active (pressed), disabled with shadows and color feedback
- Ensured WAI-ARIA compliance for accessibility

GUIDOURI committed:
feat(ui): refactor robust and adaptive Button component with modular styles

- Refactored Button component to support all usage contexts (labels, objects, JSX logs, etc.)
- Ensured compatibility with HTMLAnchorElement via `href` and improved props safety
- Added `aria-pressed` for better accessibility when `selected` is true
- Split button styling into modular CSS files: core, layouts, animations, and variant styles
- Applied BEM, OOCSS and SMACSS principles for scalability and maintainability
- Introduced `ButtonContent` and `generateButtonClassNames` helpers (optionally usable)
- Preserved compatibility with existing Storybook stories

GUIDOURI committed:
feat(ui/button): refactor component for flexibility, accessibility and full TypeScript support

- Converted Button to polymorphic component using forwardRef
- Supports dynamic tag rendering via `as` or `href` (button, anchor, div, span)
- Improved TypeScript types to satisfy strict ESLint rules
- Separated `ref` from props for compatibility with Vite + Storybook
- Added runtime development warnings for misuse (e.g. string onClick)
- Added fallback label when no icon or children are provided
- Ensures accessibility with aria-pressed and semantic structure

GUIDOURI committed:
test(ui-button): add Playwright E2E tests covering variant styles, icon, click, disabled and layout behaviors
